<a href="https://colab.research.google.com/github/aalvinsu/Alvin-Public/blob/main/h8_p0lc3_Alvin.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Data Impor

In [18]:
from google.colab import auth
from google.cloud import bigquery

# Autentikasi menggunakan JSON key file
auth.authenticate_user()
project_id = 'iconic-elevator-196204'
client = bigquery.Client(project=project_id)

# Query SQL untuk mengambil data dari BigQuery
query = """
    SELECT *
    FROM `bigquery-public-data.austin_crime.crime`
    LIMIT 10
"""

# Eksekusi query dan ambil hasilnya
query_job = client.query(query)
results = query_job.result()


Mengambil data lansgung dari BigQuery dengan autensifikasi menggunakan JSON key file, sehingga kita bisa melakukan analisis dan eksplorasi data di Google Collab

## Tampilkan tabel crime sampai 5 baris pertama

In [19]:
# Menampilkan Head
for i, row in enumerate(results):
    print(row)
    if i == 4:  # Menghentikan setelah 5 baris pertama
        break

Row((2015150483, 'Austin, TX', None, datetime.datetime(2015, 1, 27, 12, 0, tzinfo=datetime.timezone.utc), 'Not cleared', None, 'RAPE                          ', 'B', None, None, None, 'nan', 'Rape', datetime.datetime(2015, 1, 15, 12, 0, tzinfo=datetime.timezone.utc), None, None, 2015, None), {'unique_key': 0, 'address': 1, 'census_tract': 2, 'clearance_date': 3, 'clearance_status': 4, 'council_district_code': 5, 'description': 6, 'district': 7, 'latitude': 8, 'longitude': 9, 'location': 10, 'location_description': 11, 'primary_type': 12, 'timestamp': 13, 'x_coordinate': 14, 'y_coordinate': 15, 'year': 16, 'zipcode': 17})
Row((20151542027, 'Austin, TX', None, datetime.datetime(2015, 7, 24, 12, 0, tzinfo=datetime.timezone.utc), 'Not cleared', None, 'RAPE                          ', 'D', None, None, None, 'nan', 'Rape', datetime.datetime(2015, 6, 3, 12, 0, tzinfo=datetime.timezone.utc), None, None, 2015, None), {'unique_key': 0, 'address': 1, 'census_tract': 2, 'clearance_date': 3, 'clear

## Apa saja jenis tindak kejahatan yang terjadi di kota Austin selama tahun 2016?

Rape, Theft, Robbery, Agg Assault, Burglary, Auto Theft, Murder

In [20]:
# Jenis tindak kejahatan yang terjadi di Austin pada tahun 2016
query = """
    SELECT DISTINCT primary_type
    FROM `bigquery-public-data.austin_crime.crime`
    WHERE EXTRACT(YEAR FROM timestamp) = 2016
"""

# Eksekusi query dan ambil hasilnya
query_job = client.query(query)
results = query_job.result()

# Menampilkan hasil
print("Jenis Tindak Kejahatan yang Terjadi di Kota Austin pada Tahun 2016:")
for row in results:
    print(row.primary_type)

Jenis Tindak Kejahatan yang Terjadi di Kota Austin pada Tahun 2016:
Rape
Theft
Robbery
Agg Assault
Burglary
Auto Theft
Murder


## Berapa banyak kasus tindak kejahatan berdasarkan kasus pembersihannya (clearance status) selama tahun 2016?

Not cleared: 29715
Cleared by Exception: 1014
None: 1719
Cleared by Arrest: 5013

In [21]:
# Jumlah kasus tindak kejahatan berdasarkan clearance_status pada tahun 2016
query = """
    SELECT clearance_status, COUNT(*) AS total_cases
    FROM `bigquery-public-data.austin_crime.crime`
    WHERE EXTRACT(YEAR FROM timestamp) = 2016
    GROUP BY clearance_status
"""

# Eksekusi query dan ambil hasilnya
query_job = client.query(query)
results = query_job.result()

# Menampilkan hasil
print("Jumlah Kasus Tindak Kejahatan Berdasarkan Clearance Status pada Tahun 2016:")
for row in results:
    print(f"{row.clearance_status}: {row.total_cases}")

Jumlah Kasus Tindak Kejahatan Berdasarkan Clearance Status pada Tahun 2016:
Not cleared: 29715
Cleared by Exception: 1014
None: 1719
Cleared by Arrest: 5013


## Jenis kejahatan apa saja yang paling banyak belum terselesaikan kasusnya di tahun 2016?

Thefty, BUrgalary, Auto Theft, Agg Assault, Robbery

In [22]:
# Jenis tindak kejahatan yang paling banyak belum terselesaikan kasusnya pada tahun 2016
query = """
    SELECT primary_type, COUNT(*) AS total_cases
    FROM `bigquery-public-data.austin_crime.crime`
    WHERE EXTRACT(YEAR FROM timestamp) = 2016 AND clearance_status = 'Not cleared'
    GROUP BY primary_type
    ORDER BY total_cases DESC
    LIMIT 5
"""

# Eksekusi query dan ambil hasilnya
query_job = client.query(query)
results = query_job.result()

# Menampilkan hasil
print("Jenis Tindak Kejahatan yang Paling Banyak Belum Terselesaikan Kasusnya pada Tahun 2016:")
for row in results:
    print(f"{row.primary_type}: {row.total_cases}")

Jenis Tindak Kejahatan yang Paling Banyak Belum Terselesaikan Kasusnya pada Tahun 2016:
Theft: 21894
Burglary: 4440
Auto Theft: 1540
Agg Assault: 919
Robbery: 560


## **Di bulan apa di tahun 2016 kasus pencurian (theft) sangat banyak terjadi?**

Kasus pencurian (theft) sangat banyak terjadi pada bulan 4 dengan total kasus 2282

In [23]:
# Menentukan bulan di tahun 2016 ketika kasus pencurian (theft) terjadi sangat banyak
query = """
    SELECT EXTRACT(MONTH FROM timestamp) AS month, COUNT(*) AS total_cases
    FROM `bigquery-public-data.austin_crime.crime`
    WHERE EXTRACT(YEAR FROM timestamp) = 2016 AND primary_type = 'Theft'
    GROUP BY month
    ORDER BY total_cases DESC
    LIMIT 1
"""

# Eksekusi query dan ambil hasilnya
query_job = client.query(query)
results = query_job.result()

# Menampilkan hasil
for row in results:
    month = row.month
    total_cases = row.total_cases
    print(f"Kasus pencurian (theft) sangat banyak terjadi pada bulan {month} dengan total kasus {total_cases}")

Kasus pencurian (theft) sangat banyak terjadi pada bulan 4 dengan total kasus 2282


## **Di distrik apa yang paling banyak terjadi kasus pencurian selama 2016?**

Distrik dengan kasus pencurian (theft) terbanyak pada tahun 2016: D dengan total kasus 4117

In [24]:
# Menentukan distrik dengan jumlah kasus pencurian (theft) terbanyak selama tahun 2016
query = """
    SELECT district, COUNT(*) AS total_cases
    FROM `bigquery-public-data.austin_crime.crime`
    WHERE EXTRACT(YEAR FROM timestamp) = 2016 AND primary_type = 'Theft'
    GROUP BY district
    ORDER BY total_cases DESC
    LIMIT 1
"""

# Eksekusi query dan ambil hasilnya
query_job = client.query(query)
results = query_job.result()

# Menampilkan hasil
for row in results:
    district = row.district
    total_cases = row.total_cases
    print(f"Distrik dengan kasus pencurian (theft) terbanyak pada tahun 2016: {district} dengan total kasus {total_cases}")

Distrik dengan kasus pencurian (theft) terbanyak pada tahun 2016: D dengan total kasus 4117


***Berapa lama rata-rata waktu (dalam hari) kasus tindak kejahatan 'theft' terselesaikan sejak kasus tersebut terangkat selama tahun 2016?***

Rata-rata waktu penyelesaian kasus tindak kejahatan 'theft' selama tahun 2016: 17.63 hari

In [25]:
# Mnghitung rata-rata waktu (dalam hari) untuk menyelesaikan kasus tindak kejahatan 'theft'
query = """
    SELECT AVG(DATE_DIFF(clearance_date, timestamp, DAY)) AS average_duration
    FROM `bigquery-public-data.austin_crime.crime`
    WHERE EXTRACT(YEAR FROM timestamp) = 2016 AND primary_type = 'Theft'
"""

# Eksekusi query dan ambil hasilnya
query_job = client.query(query)
results = query_job.result()

# Menampilkan hasil
for row in results:
    average_duration = row.average_duration
    print(f"Rata-rata waktu penyelesaian kasus tindak kejahatan 'theft' selama tahun 2016: {average_duration:.2f} hari")

Rata-rata waktu penyelesaian kasus tindak kejahatan 'theft' selama tahun 2016: 17.63 hari


### ***Dari penjabaran yang sudah kamu bahas (pertanyaan 2-7), apa kesimpulan yang kamu dapat sampaikan?***

Selama tahun 2016 telah terjadi banyak sekali tindak kejahatan dimana biasanya tindak kejahatan yang terjadi adalah Rape, Theft, Robbery, Agg Assault, Burglary, Auto Theft, Murder. Yangmana kasus tindak kejahatan theft adalah kasus terbanyak yang belum terselesaikan dengan jumlah 21,894. Angka ini disumbang dari bulan April sebagai kasus pencurian terbanyak sejumlah 2,282 kasus.
Kasus pencurian terbanyak terjadi di distrik D dimana di distrik ini banyak orang kehilangan harta benda karena pencurian.

Sehingga rekomendasi yang bisa saya berikan adalah memperkuat keamanan pada semua distrik teruatama distrik D, ditambah agar angka ini bisa terus menurun kita bisa melakukan pencegahan dengan memasang CCTV, menerapkan double lock pada pintu.